src/style/style_layer/heatmap_style_layer.js

```diff
@@ -0,0 +1,58 @@
+// @flow
+
+const StyleLayer = require('../style_layer');
+const HeatmapBucket = require('../../data/bucket/heatmap_bucket');
+const RGBAImage = require('../../util/image').RGBAImage;
+
+import type Texture from '../../render/texture';
+
+class HeatmapStyleLayer extends StyleLayer {
+
+    heatmapTexture: ?WebGLTexture;
+    heatmapFbo: ?WebGLFramebuffer;
+    colorRampData: Uint8Array;
+    colorRamp: RGBAImage;
+    colorRampTexture: ?Texture;
+
+    createBucket(options: any) {
+        return new HeatmapBucket(options);
+    }
+
+    constructor(layer: LayerSpecification) {
+        super(layer);
+        this.colorRampData = new Uint8Array(256 * 4);
+        if (!this.getPaintProperty('heatmap-color')) {
+            this.setPaintProperty('heatmap-color', this._paintSpecifications['heatmap-color'].default, '');
+        }
+    }
+
+    _applyPaintDeclaration(name: any, declaration: any, options: any, globalOptions: any, animationLoop: any, zoomHistory: any) {
+        super._applyPaintDeclaration(name, declaration, options, globalOptions, animationLoop, zoomHistory);
+        if (name === 'heatmap-color') {
+            const len = this.colorRampData.length;
+            for (let i = 4; i < len; i += 4) {
+                const pxColor = this.getPaintValue('heatmap-color', {zoom: i / len});
+                const alpha = pxColor[3];
+                this.colorRampData[i + 0] = Math.floor(pxColor[0] * 255 / alpha);
+                this.colorRampData[i + 1] = Math.floor(pxColor[1] * 255 / alpha);
+                this.colorRampData[i + 2] = Math.floor(pxColor[2] * 255 / alpha);
+                this.colorRampData[i + 3] = Math.floor(alpha * 255);
```

Comment:
Thanks for the explanation — I'll look more closely at the premultiplying stuff.

> It'd also be cool to have unit tests for generating color ramp textures/buffers from a style specification.

I think it would be enough to cover this with a render test, since the results are visual, not something where a difference would be useful to see as numbers in the console. What do you think?

Reply:
Per chat, the code works correctly — leaving it as is as a weird special case because we need to un-premultiply the premultiplied colors returned by `getPaintValue` because the `Texture` class later premultiplies them when uploading again.